sync/sync.go

```diff
@@ -258,90 +251,114 @@ func (s *Synchronizer) handlePluginRevertBlock() {
 	}
 }
 
-//nolint:gocyclo
-func (s *Synchronizer) verifierTask(ctx context.Context, block *core.Block, stateUpdate *core.StateUpdate,
-	newClasses map[felt.Felt]core.Class, resetStreams context.CancelFunc, persisted chan struct{},
+func (s *Synchronizer) verifierTask(
+	ctx context.Context,
+	committedBlock *CommittedBlock,
+	resetStreams context.CancelFunc,
 ) stream.Callback {
 	verifyTimer := time.Now()
-	commitments, err := s.blockchain.SanityCheckNewHeight(block, stateUpdate, newClasses)
-	if err == nil {
-		s.listener.OnSyncStepDone(OpVerify, block.Number, time.Since(verifyTimer))
+	commitments, err := s.blockchain.SanityCheckNewHeight(
+		committedBlock.Block,
+		committedBlock.StateUpdate,
+		committedBlock.NewClasses,
+	)
+	if err != nil {
+		return func() {
+			defer close(committedBlock.Persisted)
+			s.log.Warnw("Sanity checks failed", "number", committedBlock.Block.Number, "hash", committedBlock.Block.Hash.ShortString(), "err", err)
+			resetStreams()
+		}
 	}
+
+	s.listener.OnSyncStepDone(OpVerify, committedBlock.Block.Number, time.Since(verifyTimer))
 	return func() {
-		select {
-		case <-ctx.Done():
-			return
-		default:
-			defer close(persisted)
+		s.storeTask(ctx, committedBlock, resetStreams, commitments)
```

Comment:
Thank you for this refactor. I believe it improves code legibility so much more 🙏